data/libs/bind-manager.lua

```diff
@@ -0,0 +1,225 @@
+--
+-- Example:
+--
+-- > my_bind = bindManager.getBindingDesc(keybinding)
+--
+-- Parameters:
+--
+--   keybinding - a table generated from InputBinding::KeyBinding, see LuaInput.cpp
+--
+-- Returns:
+--
+--   string
+--
+function bindManager.getBindingDesc(bind)
+	if not bind then return end
+	local axis_names = { linput.X, linput.Y, linput.Z }
+	local mouse_names = { linput.MOUSE_LMB, linput.MOUSE_MMB, linput.MOUSE_RMB }
+	if bind.key then
+		return Input.GetKeyName(bind.key)
+	elseif bind.joystick and bind.hat then
+		return bindManager.joyAcronym(bind.joystick) .. linput.HAT .. bind.hat .. linput.DIRECTION .. bind.dir
+	elseif bind.joystick and bind.button then
+		return bindManager.joyAcronym(bind.joystick) .. linput.BUTTON .. bind.button
+	elseif bind.joystick and bind.axis then
+		return (bind.direction < 0 and "-" or "") .. bindManager.joyAcronym(bind.joystick) .. linput.AXIS .. (axis_names[bind.axis + 1] or tostring(bind.axis))
+	elseif bind.mouse then
+		return linput.MOUSE .. (mouse_names[bind.mouse] or tostring(bind.mouse))
+	end
+end
+
+--
+-- Function: bindManager.getChordDesc
+--
+-- Get a localized name for a key chord to display on a binding
+--
+-- Example:
+--
+-- > local binding_string = bindManager.getChordDesc(action_binding)
+--
+-- Parameters:
+--
+--   action_binding - a LuaInputAction binding (see GetBindingPages)
+--
+-- Returns:
+--
+--   string, some key combination
+--
+function bindManager.getChordDesc(chord)
+	if not chord.enabled then return "" end
+
+	local str = bindManager.getBindingDesc(chord.activator)
+	local mod1 = chord.modifier1
+	local mod2 = chord.modifier2
+
+	if mod1 then str = bindManager.getBindingDesc(mod1) .. " + " .. str end
+	if mod2 then str = bindManager.getBindingDesc(mod2) .. " + " .. str end
+
+	return str
+end
+
+--
+-- Function: bindManager.localizeBindingId
+--
+-- Convert an axis or action binding style ID to a translation resource
+-- identifier, and get by this identifier a localized string from the
+-- "input-core" module.
+--
+-- Example:
+--
+-- > local tooltip = bindManager.localizeBindingId(binding_name)
+--
+-- Parameters:
+--
+--   binding_name - string, axis or action binding ID
+--
+-- Returns:
+--
+--   table:
+--     action
+--
+--   action_binding - a LuaInputAction binding (see GetBindingPages)
+--
+function bindManager.localizeBindingId(str)
+	local jsonIndex = str:gsub("([^A-Z0-9_])([A-Z0-9])", "%1_%2"):upper()
+	return rawget(linput, jsonIndex) or ('[NO_JSON] '..jsonIndex)
+end
+
+--
+-- Function: bindManager.joyAcronym
+--
+-- Returns a string with the acronym for given joystick's ID,
+--
+-- Example:
+--
+-- > ui.text("  " .. bindManager.joyAcronym(joystickID) .. ":")
+--
+-- Parameters:
+--
+--   id - number, KeyBinding.joystick
+--
+-- Returns:
+--
+--   a string, "Joy1","Joy2" etc.
+--
+function bindManager.joyAcronym(id)
+	if id == 65535 then return "DISABLED" end
+	return "Joy" .. tostring(id)
+end
```

Comment:
Indeed, after compiling and testing the game, I did not see a place where the `DISABLED` text is actually used, so this function can likely be removed.

Reply:
@Web-eWorks the fact is that this function is used in two places - in the list of joysticks (in settings-window), and in `getBindingDesc`. Removing the function will cause some code duplication, and the convention of naming the joystick "Joy" will no longer be formalized in the code.
As I can see, the `Input::JoystickFromGUID` function can return -1, which will be converted to 65535 for `uint16_t`, so without the function,  `Joy65535` will be output.  This can happen if the binding string contains a GUID that does not have a description. Most likely, this can only happen as a result of manual editing of the config.

Reply:
Alright, in this case I'd recommend amending the function to pull the format strings from the translation resource anyways. It's cleaner and allows the game to be more fully translated in languages where it makes sense (e.g. character-based languages like Chinese or Japanese).

Reply:
Done!